docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+         */
+        long averageTransferBytesPerSecond();
+
+        /**
+         * The transfer request for the object whose bytes were transferred.
+         */
+        TransferObjectRequest objectRequest();
+
+        /**
+         * The number of bytes transferred for this event.
+         */
+        long bytes();
+
+        /**
+         * The total length of the object.
+         */
+        long objectLength();
+
+        /**
+         * If the transfer of the given object is complete.
+         */
+        boolean complete();
+    }
+
+    interface Completed extends Initiated {
+    }
+
+    interface Cancelled extends Initiated {
+    }
+
+    interface Failed extends Initiated {
+        /**
+         * The error.
+         */
+        Throwable error();
+    }
+}
+
+/**
+ * Hints to help optimize an object download.
+ */
+public interface DownloadHints {
+    /**
+     * The size of the object to be downloaded. In parallel download situations
+     * this allows the Transfer Manager to avoid making an extra request to S3
+     * to find the size of the object in order to determine proper object
+     * partitioning.
+     */
+    long objectLength();
+}
+
+/**
+ * A download transfer of a single object from S3.
+ */
+public interface DownloadObject extends Transfer<DownloadObject> {
+    @Override
+    CompletableFuture<DownloadObject> completionFuture();
+
+    @Override
+    DownloadObjectState pause();
+}
+
+
+/**
+ * An upload transfer of a single object to S3.
+ */
+public interface UploadObject extends Transfer<UploadObject> {
+    @Override
+    CompletableFuture<UploadObject> completionFuture();
+
+    @Override
+    UploadObjectState pause();
+}
+
+/**
+ * Represents the transfer of one or more objects to or from S3.
+ */
+public interface Transfer<T extends Transfer> {
+
+    CompletableFuture<T> completionFuture();
+
+    /**
+     * Pause this transfer, cancelling any requests in progress.
+     * <p>
+     * The returned state object can be used to resume this transfer at a later
+     * time.
+     *
+     * @throws IllegalStateException If this transfer is completed or cancelled.
+     * @throws UnsupportedOperationException If the transfer does not support
+     * pause and resume.
+     */
+    TransferState pause();
+}
+
+/**
+ * Encapsulation of the state of the object.
+ */
+public interface TransferState {
+    /**
+     * Persist this state so it can later be resumed.
+     */
+    void persistTo(OutputStream os);
+}
+
+/**
+ * A request to download an object. The object to download may be specified as
+ * an S3 {@link GetObjectRequest} or as a presigned URL.
+ */
+public class DownloadObjectRequest extends AbstractTransferRequest {
+    /**
+     * The S3 GetObjectRequest representing the object to download.
+     */
+    public Optional<GetObjectRequest> getObjectRequest() {
+        return Optional.ofNullable(getObjectRequest);
+    }
```

Comment:
Instead of optional for these "either" fields, should we have a legit union type with is*, as* and visitor methods?

Reply:
Done. I'm not happy about the name `UploadObjectSpecification`/`DownloadObjectSpecification` but don't have a better one atm.